win32ss/user/ntuser/ghost.c

```diff
@@ -176,20 +433,64 @@ BOOL FASTCALL IntMakeHungWindowGhosted(HWND hwndHung)
 
     if (IntIsGhostWindow(pHungWnd))
     {
-        DPRINT1("IntIsGhostWindow\n");
+        ERR("IntIsGhostWindow\n");
         return FALSE;   // ghost window cannot be ghosted
     }
 
-    if (IntGhostWindowFromHungWindow(pHungWnd))
+    if (IntIsShellTrayWnd(pHungWnd))
     {
-        DPRINT("Already ghosting\n");
-        return FALSE;   // already ghosting
+        TRACE("IntIsShellTrayWnd\n");
+        return FALSE;   // Shell_TrayWnd
     }
 
-    // TODO: Find a way to pass the hwnd of pHungWnd to the ghost thread as we can't pass parameters directly
+    if (IntIsProgmanWnd(pHungWnd))
+    {
+        TRACE("IntIsProgmanWnd\n");
+        return FALSE;   // Progman
+    }
 
+    // initialize event
     if (!gptiGhostThread)
-        UserCreateSystemThread(ST_GHOST_THREAD);
+    {
+        InitializeListHead(&gGhostListHead);
+
+        // create event
+        gpGhostStartupEvent = ExAllocatePoolWithTag(NonPagedPool,
+                                                    sizeof(KEVENT),
+                                                    USERTAG_GHOST);
+        if (!gpGhostStartupEvent)
+        {
+            ERR("Unable to allocate event\n");
+            return FALSE;   // creation failed
+        }
+        KeInitializeEvent(gpGhostStartupEvent, SynchronizationEvent, FALSE);
+
+        // create ghost thread
+        if (!UserCreateSystemThread(ST_GHOST_THREAD))
+        {
+            ERR("Unable to create ghost thread for pHungWnd(%p), hwndHung(%p)\n",
+                (void *)pHungWnd, (void *)hwndHung);
+            return FALSE;   // creation failed
+        }
+    }
+
+    // wait for event
+    UserLeave();
+    Status = KeWaitForSingleObject(gpGhostStartupEvent, UserRequest, UserMode, FALSE, NULL);
+    UserEnterExclusive();
+    if (Status == STATUS_USER_APC)
+    {
+        ERR("KeWaitForSingleObject failed: Status: 0x%08lx\n", Status);
+        return FALSE;
+    }
+
+    // Pass the hwnd of pHungWnd to the ghost thread as we can't pass parameters directly
+    if (!UserPostThreadMessage(gptiGhostThread, GTM_CREATE_GHOST, (WPARAM)hwndHung, 0))
```

Comment:
Sorry, it's too difficult for this fool. Please simply order to me.

Reply:
My recommendation would be to add the window to `gGhostListHead` above before dropping the lock. You can then either wait for the thread & post a message or even just directly inform the thread via a KEVENT that it needs to create a new window (and wait for the event in the ghost thread in addition to window messages, using some equivalent of MsgWaitForMultipleObjects).
That way the ghost thread won't terminate because the list is not empty.

Reply:
Added the window before dropping the lock.
Added gpGhostDestroyEvent.